File: typescript/src/types/transaction.ts
import type { CertificateData } from './certificate.js';

/**
 * The type of transaction to build.
 *
 * - `default`        — standard certificate issuance against an existing state.
 * - `bootstrap`      — initialise a new state (fork) from a bootstrap datum.
 * - `custom`         — custom certificate transaction.
 * - `burn_state`     — burn the state token and close the workspace.
 * - `burn_bootstrap` — burn a bootstrap datum token.
 * - `init`           — deploy the initial bootstrap datum.
 * - `deploy`         — deploy a library proxy contract.
 */
export type TransactionType =
  | 'default'
  | 'bootstrap'
  | 'custom'
  | 'burn_state'
  | 'burn_bootstrap'
  | 'init'
  | 'deploy';

/** Status information returned alongside an unsigned transaction. */
export interface BuildStatus {
  message: string;
  code: number;
}

/** Request body for `POST /api/v1/transaction/build`. */
export interface BuildTransactionRequest {
  /** Array of certificates to issue in this transaction. */
  certificates: CertificateData[];
  /** Transaction type. */
  type: TransactionType;
  /** Cardano address of the signer / payer. */
  address: string;
  /**
   * Bootstrap datum reference. Required when `type` is `bootstrap`
   * or when you want to specify which bootstrap datum to fork from.
   */
  bootstrapDatum?: Record<string, unknown>;
  /** Existing state ID to update. Required for `default` and `burn_state` types. */
  stateId?: string;
}

/** Response body for `POST /api/v1/transaction/build`. */
export interface BuildTransactionResponse {
  /** CBOR-hex encoded unsigned transaction, ready to be signed. */
  unsignedTransaction: string;
  /** Echoes back the transaction type. */
  type: TransactionType;
  /** Build status with a human-readable message and code. */
  status: BuildStatus;
}

File: python/uverify_sdk/models/transaction.py
"""Transaction-related data models."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, List, Literal, Optional

from .certificate import CertificateData

TransactionType = Literal[
    "default",
    "bootstrap",
    "custom",
    "burn_state",
    "burn_bootstrap",
    "init",
    "deploy",
]


@dataclass
class BuildTransactionRequest:
    """
    Request body for ``POST /api/v1/transaction/build``.

    Attributes:
        certificates:    List of certificates to issue.
        type:            Transaction type.
        address:         Cardano address of the signer / payer.
        bootstrap_datum: Optional bootstrap datum reference.
        state_id:        Existing state ID (required for ``default`` / ``burn_state``).
    """

    certificates: List[CertificateData]
    type: TransactionType
    address: str
    bootstrap_datum: Optional[Dict[str, Any]] = None
    state_id: Optional[str] = None

    def to_dict(self) -> dict:
        d: dict = {
            "certificates": [c.to_dict() for c in self.certificates],
            "type": self.type,
            "address": self.address,
        }
        if self.bootstrap_datum is not None:
            d["bootstrapDatum"] = self.bootstrap_datum
        if self.state_id is not None:
            d["stateId"] = self.state_id
        return d


@dataclass
class BuildStatus:
    message: str
    code: int

    @classmethod
    def from_dict(cls, data: dict) -> "BuildStatus":
        return cls(message=data["message"], code=data["code"])


@dataclass
class BuildTransactionResponse:
    """
    Response from ``POST /api/v1/transaction/build``.

    Attributes:
        unsigned_transaction: CBOR-hex encoded unsigned transaction.
        type:                 Echoed back transaction type.
        status:               Build status.
    """

    unsigned_transaction: str
    type: str
    status: BuildStatus

    @classmethod
    def from_dict(cls, data: dict) -> "BuildTransactionResponse":
        return cls(
            unsigned_transaction=data["unsignedTransaction"],
            type=data["type"],
            status=BuildStatus.from_dict(data["status"]),
        )
